### ISSUE TITLE
multiline code block crashes xonsh

### ISSUE THREAD

**Issue by laerus** [xonsh/xonsh#1975]
```
$ @contextmanager
Traceback (most recent call last):
  File "/home/laerus/.local/bin/xonsh", line 6, in <module>
    exec(compile(open(__file__).read(), __file__, 'exec'))
  File "/home/laerus/src/xonsh/scripts/xonsh", line 3, in <module>
    main()
  File "/home/laerus/src/xonsh/xonsh/main.py", line 221, in main
    shell.shell.cmdloop()
  File "/home/laerus/src/xonsh/xonsh/readline_shell.py", line 395, in cmdloop
    self._cmdloop(intro=intro)
  File "/home/laerus/src/xonsh/xonsh/readline_shell.py", line 381, in _cmdloop
    stop = self.onecmd(line)
  File "/usr/lib/python3.5/cmd.py", line 211, in onecmd
    return self.default(line)
  File "/home/laerus/src/xonsh/xonsh/base_shell.py", line 308, in default
    src, code = self.push(line)
  File "/home/laerus/src/xonsh/xonsh/base_shell.py", line 385, in push
    return self.compile(src)
  File "/home/laerus/src/xonsh/xonsh/base_shell.py", line 420, in compile
    return src, code
UnboundLocalError: local variable 'code' referenced before assignment
```

bisect points to 8f5aa9d73df9701a77c0b6f272439f4a4fc76da1

**Comment by laerus** [comment]
this seems to be in any multiline code

**Comment by mitnk** [comment]
yes, I could reproduce it.
```
$ ls -lh $(history file).strip()
...............
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.5/bin/xonsh", line 3, in <module>
    main()
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/xonsh/main.py", line 221, in main
    shell.shell.cmdloop()
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/xonsh/ptk/shell.py", line 126, in cmdloop
    self.default(line)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/xonsh/base_shell.py", line 308, in default
    src, code = self.push(line)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/xonsh/base_shell.py", line 387, in push
    return self.compile(src)
  File "/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/xonsh/base_shell.py", line 422, in compile
    return src, code
UnboundLocalError: local variable 'code' referenced before assignment
```
I know, I better use `ls -lh @$(history file)` here.

**Comment by scopatz** [comment]
Should be fixed in #1983
